components/renku_data_services/notebooks/api/classes/k8s_client.py

```diff
@@ -945,47 +946,80 @@ async def delete_secret(self, name: str) -> None:
+        self._kr8s_type: type[_Kr8sType] = kr8s_type
+        self._username_label = username_label
+        self._servers: list[_SessionType] = []
+
     def sanitize(self, obj: APIObject) -> Any:
         """Sanitize a JSON object."""
         raise NotImplementedError()
 
     def namespace(self) -> str:
         """Return the current kubernetes namespace."""
-        raise NotImplementedError()
+        return "a-name-space"
 
     async def cluster_name_by_class_id(self, class_id: int | None, api_user: APIUser) -> str:
         """Retrieve the cluster name given the resource class id."""
-        raise NotImplementedError()
+        return "a-cluster"
 
     async def list_servers(self, safe_username: str) -> list[_SessionType]:
         """List all the user sessions visible from the safe_username."""
-        raise NotImplementedError()
+        return [s for s in self._servers if safe_username in s.metadata.labels.values()]
 
     async def create_server(
         self, manifest: _SessionType, api_user: AnonymousAPIUser | AuthenticatedAPIUser
     ) -> _SessionType:
         """Launch a user session."""
-        raise NotImplementedError()
+        manifest.metadata.labels[self._username_label] = api_user.id
+        manifest.metadata.creationTimestamp = datetime.datetime.fromisoformat("2025-03-05T15:24:55.474Z")
+
+        self._servers.append(manifest)
+        return manifest
 
     async def get_server(self, name: str, safe_username: str) -> _SessionType | None:
         """Lookup a user session by name."""
-        raise NotImplementedError()
+        for s in self._servers:
+            if s.metadata.name == name:
+                return s
+        return None
 
     async def get_server_logs(
         self, server_name: str, safe_username: str, max_log_lines: Optional[int] = None
     ) -> dict[str, str]:
         """Retrieve the logs for the given user session."""
-        raise NotImplementedError()
+        server = await self.get_server(server_name, safe_username)
+        if server is not None:
+            return {
+                "container-0": "fake log lines\n fake log lines",
+                "container-1": "fake log lines 2 \n fake log lines 2",
+            }
+
+        raise errors.MissingResourceError(
+            message=f"Cannot find server {server_name} for user {safe_username} to retrieve logs."
+        )
 
     async def patch_server(
         self, server_name: str, safe_username: str, patch: dict[str, Any] | list[dict[str, Any]]
     ) -> _SessionType:
-        """Patch the user session."""
-        raise NotImplementedError()
+        """Pretend to Patch the user session."""
+        server = await self.get_server(server_name, safe_username)
+        if not server:
+            raise errors.MissingResourceError(
+                message=f"Cannot find server {server_name} for user {safe_username} in order to patch it."
+            )
+        return server
```

Comment:
you could probably actually do patching using `deepmerge`

Reply:
Thanks for the review, I'll keep this in mind as I extend the tests.

I also have to think hard if I don't want to change at which level I implement a dummy class for testing. With the current implementation we can't create unit tests for most of the code within this file, which I think is a problem. 

For now I am going to proceed forward as-is and I will probably rework this quite a bit as soon as I have a demo done.